Design Tools: Add block instance elements color support for buttons and headings

aaronrobertshaw opened the pull request:
Related:
- https://github.com/WordPress/gutenberg/issues/40318

## What?

Allows styling button and heading elements on individual block instances. 

_This PR only adds button & heading color support to the Group block, other container blocks can be updated in follow-ups._

## Why?

This is a small step towards allowing section specific styling based on container block instances. More background about the long term direction and goals can be found in https://github.com/WordPress/gutenberg/issues/40318 & https://github.com/WordPress/gutenberg/issues/48581.

## How?

- Updates the color block support to add style attribute object if button or heading color support is enabled
- Updates `elements.php` to inject block instance classname if any button or heading colors have been set
- Updates `elements.php` to generate CSS styles for button and heading element styles and add them to the block-supports context
- Updates `styles.js` hook to generate CSS styles for any button and heading element styles

_Note: The approach was refactored a little to use config arrays that might more easily be extended as we look to add further elements, e.g. caption, cite, or types of styles, e.g. typography._


## Testing Instructions

1. Add a group block to a post
2. Within that group block add a range of other blocks including at least, a paragraph with a link, some buttons, and several headings of different levels
3. Select the group block and apply element styles for links, buttons, and heading through the block inspector's color panel
4. Confirm these element styles are correctly displayed within the editor
5. Save the post and confirm matching application of styles of the fronend
6. Give link hover states a test along with gradient backgrounds
    - Note: Link hover element styles currently override button text color as buttons don't yet support `:hover`
    - Note: The generic "heading" element's gradient will override a simple background color set on a specific heading element e.g. `h1` - `h6`. This is an existing issue.


## Screenshots or screencast <!-- if applicable -->

https://github.com/WordPress/gutenberg/assets/60436221/5a704729-0438-44a8-8578-a2ad96eae1e7

aaronrobertshaw committed:
Design Tools: Add block instance element styling for headings

aaronrobertshaw committed:
Add button element styling to block instances

andrewserong commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -58,6 +58,9 @@ final class WP_Style_Engine {
 					'default' => 'background-color',
 				),
 				'path'          => array( 'color', 'background' ),
+				'css_vars'      => array(
```
Just double-checked and refreshed my memory: the addition here and in `gradient` looks safe to me and doesn't affect calls from `colors.php` since in that file, we're calling `gutenberg_style_engine_get_styles` with the `convert_vars_to_classnames` option flag, which means that it skips adding the `css_vars` in cases where we intend to grab classnames instead. For example, if we add a Post Title block to a post or page and set background / gradient colors, then with this PR applied, it'll still receive classnames and not any inline styles, which is as we'd expect 👍

E.g. this is as on trunk:

<img width="1266" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/6a3e4678-6530-45c3-9e0f-4298c1d34161">

While elsewhere on the same page, the container-based element styles are applied correctly via the `background-color` property, etc:

<img width="1274" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/9948eb98-2a64-4a98-a9fb-f6de4c5d644d">

ramonjd replied:
The style engine™️ delivers

andrewserong reviewed (commented):
Just gave the style engine changes a quick smoke test — this direction is looking good to me!

aaronrobertshaw committed:
Linting fixes

aaronrobertshaw committed:
Another linting tweak

aaronrobertshaw committed:
Consolidate element style generation for blocks

If more than color styles are to be supported for elements on a per block instance basis, these refactors should make that much easier.

kevin940726 reviewed (approved):
Looks and works good to me!

andrewserong reviewed (approved):
Code-wise this is looking really good to me, too! One issue I ran into while testing, that I think is beyond the scope of this PR, is that block-level background color for a block within a Group block will not override a gradient color set further up. For example, if I set `h3` in a Group block to have a gradient, then select the Heading block and set it to have a background color:

<img width="796" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/f3b7bf7c-f428-450f-92d0-027459eb8ac1">

In the above screenshot, the heading has a blue gradient set at the Group block level for `h3` elements, and at the individual block level, I've selected a red background color. The blue gradient wins. Setting a gradient on the child block does work, however.

I think the problem seems to be that gradient colors use the `background` CSS property and that at the individual block, when we set a background color, we don't also reset `background` (or `background-image`). But doing that reset could be tricky to achieve in a reliable way, anyhow 🤔

In any case, other than that this PR's testing well for me, and the code looks good! Personally, I'd leave the issues with dealing with gradients and how they potentially conflict with other rules to another time/issue.

LGTM! ✨

aaronrobertshaw commented:
Thanks for the reviews and testing @kevin940726 and @andrewserong 🙇 

> One issue I ran into while testing, that I think is beyond the scope of this PR, is that block-level background color for a block within a Group block will not override a gradient color set further up.

This is almost the same issue as the one I noted in the test instructions where if you set a gradient on a generic `heading` element then attempt to override a specific element e.g. `h3` with a simple background color, the gradient takes precedence.  As you noted, it's due to the use of the `background` property used for gradients.

Given it's an existing issue, as it can be encountered by applying element styles for `heading` and individual heading elements on a specific block type in Global Styles, I agree that exploring background resets or alternative solutions should be a separate follow-up.

In the meantime, the worst-case scenario would be having to set a gradient that has the same color as stop points to effectively render a flat color for the background.

I'll get this merged so users have the extra flexibility. This one should prove rather valuable to pattern and theme authors.

andrewserong commented:
> I'll get this merged so users have the extra flexibility. This one should prove rather valuable to pattern and theme authors.

Absolutely, it's _very_ cool being able to use the Group block to better control the styling of child blocks in this way. Nice work getting this across the line! 🎉

aaronrobertshaw commented:
The Group block is intended only to be an initial step, so any input on which other container classes should opt into button and heading colors is more than welcome. Column/s blocks should probably get them but things like the Media & Text or Cover blocks could also conceivably adopt them.

Additionally, the long-term goal is for full support of child block-type styling, like global styles but limited to the specific section.
